### DISCUSSION TITLE
Dictating parallelism via extension

### DISCUSSION THREAD

**Discussion by mannodermaus** [junit-team/junit-framework#3476] [General]
Is it possible for a custom JUnit 5 extension to disable parallelism for the classes that it applies to?

For Android development, UI tests claim the full attention of the device, so it's impossible to run multiple of these tests in parallel (this is also enforced by Espresso, the underlying test library used for Android). Of course, I could tell users to use `ExecutionMode.SAME_THREAD` for any test class that uses my custom extension for UI tests with JUnit 5, but it would be nice if this could be dictated by the extension itself.

Thoughts? Thanks! ♥️

**Comment by sbrannen** [comment]
> Is it possible for a custom JUnit 5 extension to disable parallelism for the classes that it applies to?

That's a good question, and the short answer is "No, not an `Extension`".

But... we do have support for controlling the `ExecutionMode` for a `MethodOrderer`.

See `org.junit.jupiter.api.MethodOrderer.getDefaultExecutionMode()`.

For the sake of clarity, it is actually not possible for an `Extension` to influence the execution mode, simply because extensions are instantiated too late in the test execution lifecycle: the `ExtensionContext` is created after test execution has already started.

However, I have also wanted support for this. For example, when a Spring integration test uses `@DirtiesContext` I would like for the `SpringExtension` (or some yet-to-be-determined mechanism) to be able to disable parallel execution, and I'd like to be able to do something similar for Spring's `@RecordApplicationEvents` feature.

If we were to introduce such a feature, we'd need a way to register "something" (I'm not calling it an extension because I don't think it can be an `Extension`) that sets/changes the execution mode before test execution begins.

@marcphilipp, thoughts?

**Comment by jbduncan** [comment]
This feature request sounds quite similar to https://github.com/junit-team/junit5/issues/2677, except the linked issue would apply not just to test classes but individual tests, too.

It would simplify JUnit Pioneer's [resources injection extension](https://junit-pioneer.org/docs/resources/) a fair bit, which uses Java locks to force tests that use the same resource to run sequentially.

**Reply by mannodermaus** [reply]
Thanks @jbduncan for the hint! The `ReentrantLock` idea is brilliant and actually helps my use case a lot, at least until a proper way to dictate parallelism may be introduced in a later version of JUnit 5.

**Reply by jbduncan** [reply]
@mannodermaus You're very welcome! I'm really chuffed to hear that you like the `ReentrantLock` idea so much. The credit goes partly to @nipafx who paired with me on this extension. 😊

**Reply by sbrannen** [reply]
Indeed, @jbduncan, it is quite similar to #2677.
